text/3373-avoid-nonlocal-definitions-in-fns.md

```diff
@@ -0,0 +1,71 @@
+- Feature Name: N/A
+- Start Date: 2022-01-19
+- RFC PR: [rust-lang/rfcs#3373](https://github.com/rust-lang/rfcs/pull/3373)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Starting in Rust 2024, stop allowing items inside functions or expressions to
+implement methods or traits that are visible outside the function or
+expression.
+
+# Motivation
+[motivation]: #motivation
+
+Currently, tools cross-referencing uses and definitions (such as IDEs) must
+either search inside all function bodies and other expression-containing items
+to find potential definitions corresponding to uses within another function, or
+not cross-reference those definitions at all.
+
+Humans cross-referencing such uses and definitions may find themselves
+similarly baffled.
+
+With this change, both humans and tools can limit the scope of their search and
+avoid looking for definitions inside other functions or items, without missing
+any relevant definitions.
+
+# Explanation
+[explanation]: #explanation
+
+The following types of items, "expression-containing items", can contain
+expressions, including the definitions of other items:
+- Functions
+- Closures
+- The values assigned to `static`/`const` items
+- The discriminant values assigned to `enum` variants
+
+Starting in the Rust 2024 edition:
+- An item nested inside an expression-containing item (through any level of
+  nesting) may not define an `impl Type` block unless the `Type` is also nested
+  inside the same expression-containing item.
+- An item nested inside an expression-containing item (through any level of
+  nesting) may not define an `impl Trait for Type` unless either the `Trait` or
+  the `Type` is also nested inside the same expression-containing item.
```

Comment:
What exactly do you mean by "from coherence point of view"?  I presume you mean at least following the Orphan Rules, so that I can `impl ForeignTrait<LocalType> for ForeignType`.  But what about [crate local negative reasoning?](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=56ca62b8ce12eb1ef9e2386b67b57c15)

Reply:
@vlad20012 I'd like to be *specific* about the disallowed cases. If there are any concrete cases that the existing language doesn't cover, please give examples of those and I can expand the text accordingly.